### ISSUE TITLE
Gamepads are comfortable

### ISSUE THREAD

**Issue by es20490446e** [mupen64plus/mupen64plus-core#187]
As reported in [Launchpad](https://bugs.launchpad.net/ubuntu/+source/mupen64plus-input-sdl/+bug/1624463):

Using a generic PlayStation like game pad [Logitech Dual Action 2](http://www.kustompcs.co.uk/acatalog/4707.jpg), the automatic configuration has major usability drawbacks:

- The analog range is too short, making movement hard. In Super Smash Bros, characters don't jump diagonally.

- C-buttons, instead of isolated as in the original console game pad, are placed in the second analog. This makes pressing buttons by accident a frequent issue. In Golden Eye camera movement prevents the game from been playable, where in Road Rash you frequently discard weapons permanently.

- Z-trigger and Left-trigger are placed both in the left triggers of the game pad. But you usually need to press both simultaneously, where that placement only allows pressing one button at a time.

The solution is this:

- Ignore the console's native analog range, and choose a wider one.
- Place C-buttons in the game pad triggers.
- Place the Z-trigger in the regular button's side, where pressing the rest is usable.

After testing thoroughly various configurations with multiple games, the following seems to be the one that responds better in any circumstance:

```
# The minimum absolute value of the SDL analog joystick axis to move the N64 controller axis value from 0. For X, Y axes.
AnalogDeadzone = "338,338"
# An absolute value of the SDL joystick axis >= AnalogPeak will saturate the N64 controller axis value (at 80). For X, Y axes. For each axis, this must be greater than the corresponding AnalogDeadzone value
AnalogPeak = "21200,21200"
# Digital button configuration mappings
DPad R = "axis(4+) hat(0 Right)"
DPad L = "axis(4-) hat(0 Left)"
DPad D = "axis(5+) hat(0 Down)"
DPad U = "axis(5-) hat(0 Up)"
Start = "button(9)"
Z Trig = "button(2)"
B Button = "button(0)"
A Button = "button(1)"
C Button R = "button(5)"
C Button L = "button(4)"
C Button D = "button(7)"
C Button U = "button(6)"
R Trig = "button(3)"
L Trig = "button(8)"
Mempak switch = ""
Rumblepak switch = ""
# Analog axis configuration mappings
X Axis = "axis(0-,0+)"
Y Axis = "axis(1-,1+)"
```

**Comment by rlabrecque** [comment]
Almost all of the dual analog gamepads in m64p behave like this by default primarily because it's most similar to how it was accomplished on the Gamecube.

**Comment by es20490446e** [comment]
In PC how many people own those, compared with regular ones?

You can figure it out just looking at the number of reviews in Amazon on [GameCube game-pads](https://www.amazon.com/s/ref=sr_pg_2?rh=i%3Aaps%2Ck%3Agamecube+gamepad&page=2&keywords=gamecube+gamepad&ie=UTF8&qid=1474496044) versus [other game-pads](https://www.amazon.com/gp/search/ref=sr_pg_1?fst=as%3Aon&rh=n%3A468642%2Ck%3Agamepad&keywords=gamepad&ie=UTF8&qid=1474495842&spIA=B0171JGYF2,B01HG9Q5AG,B019FBWSWW,B01IVNBB00,B01HCRL9US,B01KQ9FAEK).

Perhaps Mupen could differentiate Nintendo game-pads from the rest.

**Comment by richard42** [comment]
I originally designed the layout of the logitech controllers (I have a bunch of them) and I chose this layout because it is the most similar in terms of button layout to that of the N64 controller.  I don't think there's any "ideal" layout though because the controllers are different, and what works well for some games might not work at all for others.  I do have some questions about your proposed layout, though.

Which games require pressing both the left trigger and the Z trigger simultaneously?  I don't understand how this could work, because I always thought the original N64 controller was designed such that your left hand would either be on the left stick (allowing you to press the left trigger and D-pad) or on the center stick (allowing you to press the Z-trigger and use the analog stick).  I thought your right hand would always be on the right stick (with the right trigger and A/B/C buttons).  With this layout, it would be impossible to simultaneously press the Z and Left triggers even with the N64 controller.

With your config, you put the left trigger on the small rectangular button on the top left, which makes it impossible to push the left trigger and use the D-pad at the same time.  I don't know if there are any games for which this would pose a problem, but it is a potential issue.

**Comment by es20490446e** [comment]
> Which games require pressing both the left trigger and the Z trigger simultaneously?

I thought it was Road Rash 64, but now that you mention it this button only serves the purpose of taunting. This also happens in Super Smash Bros, and in the rest of games I tested it also served a very secondary purpose.

> You put the left trigger on the small rectangular button on the top left, which makes it impossible to push the left trigger and use the D-pad at the same time. I don't know if there are any games for which this would pose a problem.

Most games that respond to the digital pad, they also do to the analogue one. So they assume the left trigger would be inconvenient to press for those people favouring the joystick.

But if you want to be completely sure, I can test this with a wide range of games if you wish.

**Comment by es20490446e** [comment]
If you want me to test something, let me know.

**Comment by richard42** [comment]
I just pushed a change which applies limits to the joystick coordinates in a similar shape as the controller case enforces physically on a real N64 controller. I also put the program I write to find the curves and a screenshot showing the graph into the tools folder of the core module. Can you please test the new code (it's in the input plugin) with some games (at least SSB) and report your results.  I'm concerned that this may bring further problems (you said you couldn't jump diagonally) because it restricts the X,Y joystick values allowed. But I believe that it is more accurate.  You should also consider decreasing your AnalogPeak values in the input configuration, because this will increase the sensitivity.

**Comment by es20490446e** [comment]
Okay, I'll test it.

On the other hand realise that the root cause of this problem is that the real N64 controller was very tatty, and probably trying to replicate it isn't a good idea.

**Comment by es20490446e** [comment]
I have just tested you patch with the previous games, using the default ranges in the configuration file, and I can say it greatly improves the analogue behaviour.

The problem with the jumping is gone, and the control is more precise than ever. Where before it was difficult to point in first person games, now it's pretty convenient. Also the ranges seems good to the two racing games I have tested.

I have even asked a long time user of the emulator for his opinion. And the first thing he said, without telling nothing to him, has been the control is more precise. When I asked if he found some limitation on the character speed of movement, he said he didn't.

So thank you 🕊
